jgrapht-core/src/main/java/org/jgrapht/graph/AsSynchronizedGraph.java

```diff
@@ -0,0 +1,1165 @@
+/*
+ * (C) Copyright 2018-2018, by CHEN Kui and Contributors.
+ *
+ * JGraphT : a free Java graph-theory library
+ *
+ * This program and the accompanying materials are dual-licensed under
+ * either
+ *
+ * (a) the terms of the GNU Lesser General Public License version 2.1
+ * as published by the Free Software Foundation, or (at your option) any
+ * later version.
+ *
+ * or (per the licensee's choosing)
+ *
+ * (b) the terms of the Eclipse Public License v1.0 as published by
+ * the Eclipse Foundation.
+ */
+package org.jgrapht.graph;
+
+import java.io.*;
+import java.lang.*;
+import java.util.*;
+import java.util.function.*;
+import java.util.stream.*;
+import org.jgrapht.*;
+
+/**
+ * Create a synchronized (thread-safe) Graph backed by the specified Graph. In order to guarantee
+ * serial access, it is critical that <strong>all</strong> access to the backing Graph is
+ * accomplished through the created Graph.
+ *
+ * <p>
+ * Users need to manually synchronize on {@link EdgeFactory} if creating an edge needs to access
+ * shared resources. Failure to follow this advice may result in non-deterministic behavior.
+ * </p>
+ *
+ * <p>
+ * For all methods returning a Set, the Graph guarantees that all operations on the returned Set do
+ * not affect the backing Graph. For <code>edgeSet</code> and <code>vertexSet</code> methods, the
+ * returned Set is backed by the underlying graph, but when a traversal over the set is started via
+ * a method such as iterator(), a snapshot of the underlying Set is copied for iteration purposes.
+ * For <code>edgesOf</code>, <code>incomingEdgesOf</code> and <code>outgoingEdgesOf</code> methods,
+ * the returned Set is a unmodifiable copy of the result produced by the underlying Graph. Users can
+ * control whether those copies should be cached; caching may significantly increase memory
+ * requirements. If users decide to cache those copies and the backing graph's changes don't affect
+ * them, those copies will be returned the next time the method is called. If the backing graph's
+ * changes affect them, they will be removed from cache and re-created the next time the method is
+ * called. If users decide to not cache those copies, the graph will create ephemeral copies every
+ * time the method is called. For other methods returning a Set, the Set is just the backing Graph's
+ * return.
+ * </p>
+
```

Comment:
Thanks, the code example looks good.  

For the problematic sentence, I suggest

"Even though this graph implementation is thread-safe, callers should still be aware of potential hazards from removal methods.  If calling code obtains a reference to a vertex or edge from the graph, and then calls another graph method to access information about that object, an {@link IllegalArgumentException} may be thrown if another thread has concurrently removed that object.  Therefore..."